### PR TITLE
MAILBOX-364 Part 1: Migrate EventDelivery to Reactor

### PR DESCRIPTION
**Pull request by chibenwa**
EventDelivery responsibility is to deliver an event to a collection of listeners.

Note that:
 - The "kind" of delivery wanted (synchronous / asynchronous / mixed) is now asked by the caller
 - This allow keeping a single implementation - great simplification
 - The **mixed** behaviour is now enforced - Spring & Guice
 - As reactor is used configuring the pool size no longer makes sens

### CODE REVIEW

File: mailbox/event/event-memory/pom.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            <artifactId>junit-platform-launcher</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.vintage</groupId>
```
Comment (mbaechler) [inline mailbox/event/event-memory/pom.xml]: ordering issue

File: mailbox/pom.xml [hunk 2]
```diff
@@ -61,6 +61,7 @@
         <module>tools/indexer</module>
         <module>tools/jpa-migrator</module>
         <module>tools/maildir-utils</module>
+        <module>event/event-memory</module>
```
Comment (mbaechler) [inline mailbox/pom.xml]: ordering issue

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml [hunk 3]
```diff
@@ -19,8 +19,6 @@
  -->
 
 <listeners>
-  <poolSize>8</poolSize>
```
Comment (mbaechler) [inline dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml]: no more needed, until we are in trouble again and want to tweak things (:

File: src/site/xdoc/server/config-listeners.xml [hunk 4]
```diff
@@ -44,8 +44,7 @@
             <p>
                 Some MailboxListener allows you to specify if you want to run them synchronously or asynchronously. To do so,
                 for MailboxListener that supports this, you can use the <b>async</b> attribute (optional, per mailet default) to govern the execution mode.
-                If <b>true</b> the execution will be scheduled by a FixedThreadPool whose size is defined by the <b>poolSize</b>
-                attribute (optional, default to 8). If <b>false</b>, the execution is synchronous, on the current thread.
+                If <b>true</b> the execution will be scheduled in a reactor elastic scheduler. If <b>false</b>, the execution is synchronous, on the current thread.
```
Comment (mbaechler) [inline src/site/xdoc/server/config-listeners.xml]: I don't really agree: sync does not means "same thread". Why wouldn't you want to do concurrent listener execution + join ?

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java [hunk 5]
```diff
@@ -28,28 +29,62 @@
...
+import reactor.core.scheduler.Schedulers;
+
+public class EventDeliveryImpl implements EventDelivery {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventDeliveryImpl.class);
 
     private final MetricFactory metricFactory;
 
-    @Inject
-    public SynchronousEventDelivery(MetricFactory metricFactory) {
+    public EventDeliveryImpl(MetricFactory metricFactory) {
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java]: s/EventDeliveryImpl/InVMEventDelivery/ ?

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java [hunk 6]
```diff
@@ -28,28 +29,62 @@
...
+
+    private ImmutableList<MailboxListener> filterByExecutionMode(Collection<MailboxListener> mailboxListeners, MailboxListener.ExecutionMode executionMode) {
+        return mailboxListeners.stream()
+            .filter(listener -> listener.getExecutionMode() == executionMode)
+            .collect(Guavate.toImmutableList());
+    }
+
+    private CompletableFuture<Void> doDeliver(Collection<MailboxListener> mailboxListeners, Event event) {
+        return Flux.fromIterable(mailboxListeners)
+            .flatMap(mailboxListener -> Mono.fromCallable(() ->  {
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java]: what prevents you from using ParallelFlux ?

Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java]: fromRunnable ?

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java [hunk 7]
```diff
@@ -19,11 +19,33 @@
...
 import org.apache.james.mailbox.Event;
 import org.apache.james.mailbox.MailboxListener;
 
 public interface EventDelivery {
+    class ExecutionStages {
+        private final CompletableFuture<Void> synchronousListenerFuture;
+        private final CompletableFuture<Void> asynchronousListenerFuture;
+
+        ExecutionStages(CompletableFuture<Void> synchronousListenerFuture, CompletableFuture<Void> asynchronousListenerFuture) {
+            this.synchronousListenerFuture = synchronousListenerFuture;
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: why don't you expose Mono instead ?

File: mailbox/event/event-memory/pom.xml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    <parent>
+        <artifactId>apache-james-mailbox</artifactId>
+        <groupId>org.apache.james</groupId>
+        <version>3.3.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>apache-james-mailbox-event-memory</artifactId>
+    <name>Apache James :: Mailbox :: Event :: Memory implementation</name>
```
Comment (mbaechler) [inline mailbox/event/event-memory/pom.xml]: InVM probably makes more sense for an eventbus

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml (hunk 3 above)
Comment (chibenwa) [reply dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml]: What's you plan regarding this?

Shouldn't it be more "Tweaking Reactor elastic scheduler" ?

File: src/site/xdoc/server/config-listeners.xml (hunk 4 above)
Comment (chibenwa) [reply src/site/xdoc/server/config-listeners.xml]: I will remove the 'on the current thread' stuff as it is clearly not the case

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: The big deal is that with the mono the computation did not started yet.

I don't believe the caller should be responsible of starting these computations....

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java (hunk 6 above)
Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java]: I thought flatMap was doing this for me?

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml (hunk 3 above)
Comment (mbaechler) [reply dockerfiles/run/guice/cassandra-ldap/destination/conf/listeners.xml]: I don't know for now, we can reintroduce it later, I don't have problem with that, I'm just wondering if we are not overly optimistic

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java (hunk 6 above)
Comment (mbaechler) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDeliveryImpl.java]: In theory, yes. I don't know exactly if there's a difference. Let's keep your code.

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java (hunk 7 above)
Comment (mbaechler) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: just call `.cache()` then

Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: Yes `.cache()` combined with `.subscribe()` for triggering the initial computation...

Comment (mbaechler) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: No, cache is enough, it starts the computation

Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: I do believe you but I also do have some tests failures if I do not call `.subscribe()`...

Comment (mbaechler) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: Ok at least it's what the doc says : https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#cache--

But looking at the picture, it looks like the doc is wrong, trust your test :+1:

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java [hunk 9]
```diff
@@ -88,30 +91,28 @@ public void removeGlobalListener(MailboxListener listener, MailboxSession sessio
...
+        ImmutableList<MailboxListener> listeners = ImmutableList.<MailboxListener>builder()
+            .addAll(registry.getGlobalListeners())
+            .addAll(registeredMailboxListeners(event))
+            .build();
+
+        eventDelivery.deliver(listeners, event)
+            .synchronousListenerFuture()
+            .block();
+
+        if (event instanceof MailboxDeletion) {
```
Comment (trantienduchn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java]: why did you remove this condition `listenerSnapshot.size() > 0`? I see it in old version.

File: mailbox/event/event-memory/src/test/java/org/apache/james/mailbox/events/delivery/InVmEventDeliveryTest.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    void deliverShouldHaveCalledSynchronousListenersWhenSynchronousListenerExecutedJoined() {
+        when(listener.getExecutionMode()).thenReturn(MailboxListener.ExecutionMode.SYNCHRONOUS);
+
+        inVmEventDelivery.deliver(ImmutableList.of(listener), event).synchronousListenerFuture().block();
+
+        verify(listener).event(event);
+    }
+
+    @Test
+    void deliverShouldNotBlockObAsynchronousListenersWhenSynchronousListenerExecutedJoined() {
```
Comment (trantienduchn) [inline mailbox/event/event-memory/src/test/java/org/apache/james/mailbox/events/delivery/InVmEventDeliveryTest.java]: s/deliverShouldNotBlock**Ob**/deliverShouldNotBlock**On**

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java (hunk 9 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java]: This seems like a useless optimisation: the cost of removing listeners is low if they do not exists.

I don't want to make the code more complicated for such kind of low-level optimisation. The simpler the conditions are, the better.

Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java]: Ps: Proof that it was useless: I did not broke any tests when removing it :-) #TDD_Refactoring

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java [hunk 11]
```diff
@@ -17,36 +17,36 @@
...
-            asynchronousEventDelivery.deliver(mailboxListener, event);
+        public Mono<Void> synchronousListenerFuture() {
+            return synchronousListenerFuture;
         }
-    }
 
-    @PreDestroy
-    public void stop() {
-        asynchronousEventDelivery.stop();
+        public Mono<Void> allListenerFuture() {
```
Comment (trantienduchn) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: I'm just wondering where is this method is suppose to used in our code (not test) ?

Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java]: It is tested. See InVmEventDeliveryTest.java `deliverShouldHaveCalledSynchronousListenersWhenAllListenerExecutedJoined` & `deliverShouldHaveCalledAsynchronousListenersWhenAllListenerExecutedJoined`

The idea is to keep the capabilities to await all listener (CF SynchronousEventDelivery), await none (AsyncEventDelivery) or just some of them (MixedEventDelivery) with a single implem.

Then choosing the execution model is a caller choice and not an implementation detail.

File: mailbox/event/event-memory/pom.xml [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    <parent>
+        <artifactId>apache-james-mailbox</artifactId>
+        <groupId>org.apache.james</groupId>
+        <version>3.3.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>apache-james-mailbox-event-memory</artifactId>
+    <name>Apache James :: Mailbox :: Event :: In VM implementation</name>
+    <description>Memory implementation for the eventbus API</description>
```
Comment (mbaechler) [inline mailbox/event/event-memory/pom.xml]: s/Memory/In VM

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/VoidMarker.java [hunk 13]
```diff
@@ -17,13 +17,9 @@
...
-import org.apache.james.mailbox.Event;
-import org.apache.james.mailbox.MailboxListener;
-
-public interface EventDelivery {
-
-    void deliver(MailboxListener mailboxListener, Event event);
+package org.apache.james.mailbox.events.delivery;
 
+public interface VoidMarker {
+    VoidMarker IMPL = () -> null;
```
Comment (mbaechler) [inline mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/VoidMarker.java]: not used ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java [hunk 14]
```diff
@@ -88,30 +91,28 @@ public void removeGlobalListener(MailboxListener listener, MailboxSession sessio
...
-            mailboxEvent((MailboxEvent) event);
-        }
-    }
-
-    private void mailboxEvent(MailboxEvent mailboxEvent) {
-        Collection<MailboxListener> listenerSnapshot = registry.getLocalMailboxListeners(mailboxEvent.getMailboxId());
-        if (mailboxEvent instanceof MailboxDeletion && listenerSnapshot.size() > 0) {
-            registry.deleteRegistryFor(mailboxEvent.getMailboxId());
+        ImmutableList<MailboxListener> listeners = ImmutableList.<MailboxListener>builder()
+            .addAll(registry.getGlobalListeners())
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/event/DefaultDelegatingMailboxListener.java]: we don't need to optimize it because it's probably very cheap but the right datastructure for this kind of situation would be "persistent datastructure"

File: mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java [hunk 15]
```diff
@@ -40,6 +42,7 @@
 
     private final MetricFactory metricFactory;
 
+    @Inject
     public InVmEventDelivery(MetricFactory metricFactory) {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java]: Used in DefaultDelegatingMailboxListener to expose a sinple constructor for testing.

I will add a @VisibleFortesting annotation here

**Comment by chibenwa** [comment]
`SetMessagesMethodTest.setMessagesShouldTriggerMaxQuotaReachedWhenTryingToSendMessageAndQuotaReached » Completion`

test this please

**Comment by chibenwa** [comment]
Merged
